### PR TITLE
Various redemption rate and redeem tokens improvements

### LINKED ISSUE #251: Disable redeem when bondingCurve is 0
When bondingCurve is 0, close to 0 ETH will be returned by burning tokens. By disabling the redeem button in this case, we'll give projects a tool to disable redemptions, and protect users who may not realize what they're doing from burning their tokens.

In this case, there should be a clear message explaining to the user why the button is disabled.

### LINKED ISSUE #652: "Redeem tokens for ETH" modal: Show validation error if user enters a value too large
## Description

The following comment from @johnnyd-eth should be addressed:

> In the redeem modal I've noticed there's no limit on the 'Redeem tokens for ETH' input. Maybe we should make input go red when total token supply is exceeded, and replace the 'You will receive minimum 0 ETH' for an error message?

## Proposal

We should show a real-time validation error if the maximum redeem amount is exceeded (in other words, if the user doesn't have enough tokens for the input value).

## Screenshots

example of status quo:

### LINKED ISSUE #1466: Redemption rate disabled when `distributionLimit` is 0
## Summary

Screen Shot 2022-07-21 at 3 49 23 pm

## Environment

- **Device/OS**: [e.g. iOS]
- **Browser:** [e.g. chrome, safari]
- **Hardware wallet**: [e.g. Ledger w/ Metamask]

### LINKED ISSUE #1523: Change user-facing `Bonding curve rate`s in V1 to `Redemption rate`s
## Summary

_A clear and concise description of what the bug is._

## Environment

- **Device/OS**: [e.g. iOS]
- **Browser:** [e.g. chrome, safari]
- **Hardware wallet**: [e.g. Ledger w/ Metamask]

### PR DESCRIPTION
**Pull request by johnnyd-eth**
## What does this PR do and why?

Sorry could have split this one up a bit more, but most points are linked. 

- Fixes bug where redemption rate was disabled with distribution limit is **Zero** (Closes #1466)

- Disables redeem button (shows 'Burn' button) when redemption rate is **Zero** (Closes #251)
Screen Shot 2022-07-25 at 3 05 47 pm

- Changes all V1 `Bonding curve rate`s to `Redemption rate`s (Closes #1523)

- Various redeem modal improvements (Closes #652):
     - Show when total token supply exceeded
     - Hide 'You will receive 0 ETH'
     - Change 'will' to 'would' when personal token balance exceeded
     - Make validation actually work

https://user-images.githubusercontent.com/96150256/180712644-de9d41f3-477b-434d-8723-2bcd45abcee2.mp4

## Screenshots or screen recordings

_If applicable, provide screenshots or screen recordings to demonstrate the changes._

## Acceptance checklist

- [ ] I have evaluated the [Approval Guidelines](https://github.com/jbx-protocol/juice-interface/blob/main/CONTRIBUTING.md#approval-guidelines) for this PR.
- [ ] I have tested this PR in [all supported browsers](https://github.com/jbx-protocol/juice-interface/blob/main/CONTRIBUTING.md#supported-browsers).
- [ ] I have tested this PR in dark mode and light mode (if applicable).

### CODE REVIEW

File: src/components/v2/V2Project/V2ManageTokensSection/V2RedeemModal.tsx [hunk 1]
```diff
@@ -104,7 +104,10 @@ export default function V2RedeemModal({
     if (redeemBN.eq(0)) {
       return Promise.reject(t`Required`)
     } else if (redeemBN.gt(totalBalance ?? 0)) {
-      return Promise.reject(t`Balance exceeded`)
+      return Promise.reject(t`Your balance exceeded`)
```
Comment (tomquirk) [inline src/components/v2/V2Project/V2ManageTokensSection/V2RedeemModal.tsx:107]:
```suggestion
      return Promise.reject(t`Insufficient token balance`)
```

File: src/components/v2/V2Project/V2ManageTokensSection/V2RedeemModal.tsx [hunk 2]
```diff
@@ -244,15 +255,47 @@ export default function V2RedeemModal({
...
+                      <Trans>
+                        You will receive {minReturnedTokensFormatted} ETH
+                      </Trans>
+                    )}
+                  </>
+                ) : (
+                  <>
+                    {inUSD ? (
+                      <Trans>
+                        You would receive minimum {minReturnedTokensFormatted}{' '}
```
Comment (tomquirk) [inline src/components/v2/V2Project/V2ManageTokensSection/V2RedeemModal.tsx:288]:
```suggestion
                        You would receive at least {minReturnedTokensFormatted}{' '}
```

File: src/components/v2/V2Project/V2ManageTokensSection/V2RedeemModal.tsx [hunk 3]
```diff
@@ -244,15 +255,47 @@ export default function V2RedeemModal({
...
-                </Trans>
-              ) : (
-                <Trans>You will receive {minReturnedTokensFormatted} ETH</Trans>
-              )}
+              <>
+                {!personalBalanceExceeded ? (
+                  <>
+                    {inUSD ? (
+                      <Trans>
+                        You will receive minimum {minReturnedTokensFormatted}{' '}
```
Comment (tomquirk) [inline src/components/v2/V2Project/V2ManageTokensSection/V2RedeemModal.tsx:275]:
```suggestion
                        You will receive at least {minReturnedTokensFormatted}{' '}
```

File: src/utils/v2/fundingCycle.ts [hunk 4]
```diff
@@ -23,11 +23,13 @@ import { MAX_DISTRIBUTION_LIMIT } from './math'
 export const hasDistributionLimit = (
   fundAccessConstraint: SerializedV2FundAccessConstraint | undefined,
 ): boolean => {
+  // Distribution limit defaults to Zero (which is a distribution limit)
+  if (!fundAccessConstraint) return true
```
Comment (tomquirk) [inline src/utils/v2/fundingCycle.ts:27]: **praise** nice catch.

File: src/components/v2/V2Project/V2ManageTokensSection/V2RedeemModal.tsx [hunk 5]
```diff
@@ -244,15 +255,47 @@ export default function V2RedeemModal({
...
+          minReturnedTokens?.gt(0) ? (
             <div style={{ fontWeight: 500, marginTop: 20 }}>
-              {distributionLimitCurrency?.eq(V2_CURRENCY_USD) ? (
-                <Trans>
-                  You will receive minimum {minReturnedTokensFormatted} ETH
-                </Trans>
-              ) : (
-                <Trans>You will receive {minReturnedTokensFormatted} ETH</Trans>
-              )}
+              <>
```
Comment (tomquirk) [inline src/components/v2/V2Project/V2ManageTokensSection/V2RedeemModal.tsx:270]: **suggestion** this whole block is hard to follow, and there's a bit of duplication.

Could you find a way to clean it up? Happy to offer a patch tomorrow if I'm being too vague.

File: src/components/inputs/FormattedNumberInput.tsx [hunk 6]
```diff
@@ -54,63 +56,65 @@ export default function FormattedNumberInput({
   }
 
   return (
-    <div className="formatted-number-input">
-      <Form.Item {...formItemProps}>
+    <Form.Item
```
Comment (tomquirk) [inline src/components/inputs/FormattedNumberInput.tsx:59]: **question** are you confident this doesn't break any other uses across the site? 😉 

(i havent checked)

**Review by tomquirk — CHANGES_REQUESTED** [review]
@johnnyd-eth nice.

Definitely consider breaking up these PRs in the future. We could have merged 90% of this stuff already, but now the whole thing is blocked by minor comments.

File: src/components/inputs/FormattedNumberInput.tsx (hunk 6 above)
Comment (johnnyd-eth) [reply src/components/inputs/FormattedNumberInput.tsx:59]: Yes had a look around, couldn't spot anything. Don't see how it this change could affect anything user-facing.

File: src/components/v2/V2Project/V2ManageTokensSection/V2RedeemModal.tsx (hunk 5 above)
Comment (johnnyd-eth) [reply src/components/v2/V2Project/V2ManageTokensSection/V2RedeemModal.tsx:270]: Can understand your point, however, since we want these translated, I don't think there's any way to avoid writing out 4 blocks like this. I've added a comment to clarify, but not sure there's anything else I can do

**Comment by johnnyd-eth** [comment]
Yep point on breaking this up more taken 100%. Addressed your minor comments @tomquirk

File: src/components/inputs/FormattedNumberInput.tsx (hunk 6 above)
Comment (tomquirk) [reply src/components/inputs/FormattedNumberInput.tsx:59]: Awesome sounds good.

Removing a wrapping div could definitely break stuff but lgtm.

**Review by tomquirk — APPROVED** [review]
Thanks @johnnyd-eth lgtm!
